internal/pubsub/pubsub_test.go

```diff
@@ -227,6 +228,52 @@ func TestCancelReceive(t *testing.T) {
 	}
 }
 
+type blockingDriverSub struct {
+	driver.Subscription
+	inReceiveBatch chan int
+}
+
+func (b blockingDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	b.inReceiveBatch <- 0
+	<-ctx.Done()
+	return nil, ctx.Err()
+}
+
+func TestCancelTwoReceives(t *testing.T) {
+	// We want to create the following situation:
+	// 1. Goroutine 1 calls Receive, obtains the lock (Subscription.mu),
+	//    and calls driver.ReceiveBatch, which hangs forever.
+	// 2. Goroutine 2 calls Receive.
+	// 3. The context passed to the Goroutine 2 call is canceled.
+	// We expect Goroutine 2's Receive to exit immediately. That won't
```

Comment:
Maybe I'm misinterpreting this, but it looks to me like actually goroutine 1 is the one that gets cancelled since it sends to inReceiveBatch and that gets received in the top goroutine of the test, so goroutine1 would be blocked on `<-ctx.Done()`, but goroutine2 would be blocked on `b.inReceiveBatch <- 0`.

Reply:
It's impossible to cancel goroutine 1 because the context passed to Receive is `context.Background`.

Reply:
Ah of course you're right. I still don't see how the second goroutine gets to the `<-ctx.Done()` line in `ReceiveBatch` since there isn't a second pull from `inReceiveBatch`.

Reply:
That's a good point. It's probably returning from `retry.Call`'s context check. That's probably flaky, so just to be safe, I added a buffer to that channel.